incubator/hnc/pkg/reconcilers/hierarchy_config.go

```diff
@@ -251,38 +251,77 @@ func (r *HierarchyConfigReconciler) markExisting(log logr.Logger, ns *forest.Nam
+		log.Info("Couldn't update parent", "reason", err, "parent", inst.Spec.Parent)
+		ns.SetLocalCondition(api.CritParentInvalid, err.Error())
+		return
+	}
+
+	// We now need to update the old and new parents, as well as all descendants of this namespace
+	// (since they need to update their tree labels and possibly other data).
+	if oldParent != nil {
+		r.enqueueAffected(log, "removed as parent", oldParent.Name())
+	}
+	if curParent != nil {
+		r.enqueueAffected(log, "set as parent", curParent.Name())
+	}
+	r.enqueueAffected(log, "subtree parent has changed", ns.DescendantNames()...)
+
+	// Get rid of any obsolete object conditions (issue #328)
+	r.flushObsoleteObjectConditions(log, ns, oldParent)
+}
+
+// flushObsoleteObjectConditions looks for object conditions from objects that were in one of the
+// namespaces that's been removed as a descendant, and enqueues any affected namespaces.
+func (r *HierarchyConfigReconciler) flushObsoleteObjectConditions(log logr.Logger, ns, oldAns *forest.Namespace) {
+	// Create the set of all relevant names
+	wasDesc := map[string]bool{ns.Name(): true}
+	for _, desc := range ns.DescendantNames() {
+		wasDesc[desc] = true
+	}
 
-		// Also update all descendants so they can update their labels (and in future, annotations) if
-		// necessary.
-		r.enqueueAffected(log, "subtree parent has changed", ns.DescendantNames()...)
+	// Remove conditions referring to this namespace and its descendants from all old ancestors.
+	for oldAns != nil {
+		if oldAns.ClearConditionsByNamespace(log, wasDesc) {
+			r.enqueueAffected(log, "cleared obsolete conditions", oldAns.Name())
+		}
+		oldAns = oldAns.Parent()
 	}
+
+	// In addition, this namespace may have conditions caused by a propagated object from another
+	// namespace that we need to clear. For example, we may not have been able to copy a RoleBinding
+	// from a parent into this namespace because we didn't have permission, and we've put a condition
+	// on this namespace as a result. Now that the parent has changed, we'll no longer attempt to copy
+	// that object - but we'll never clear its condition, because it's not in the list of objects seen
+	// by the object reconciler's enqueueLocalObjects.
+	//
+	// As of March 2020, the only known case like this is the CannodUpdate code, set by the object
+	// reconciler. Since we're about to resync all objects in this namespace anyway - which will
+	// result in their conditions being re-added to this namespace - it's safe to blow them all away
+	// here and just let the ones that are still applicable get reapplied.
+	//
+	// In theory, I wouldn't have expected that this would be necessary. If there was an error that
+	// caused this condition to be set, we should have returned an error to controller-runtime, which
+	// would have caused the object to be re-enqueued for reconciliation. However, when I actually
+	// watched the logs carefully, this didn't appear to happen, even after 30s. I'm not sure why.
```

Comment:
Ah, because when we failed to update `child/superadmin`, we returned an error to controller-runtime. That _should_ have re-enqueued `child/superadmin` at least one last time after `child` became a root. That last reconciliation should have realized there was nothing to do and cleared the condition. But for some reason, I didn't see `child/superadmin` get re-enqueued.

Does that make sense?

Reply:
> HNC tries to copy superadmin into child, but it fails and puts a CannotUpdate condition on child pointing to child/superadmin (which, remember, wasn't created), as well as a CannotPropagate condition on parent, also pointing to child/superadmin.

Is there an argument to made that to avoid this complexity we just don't add the `CannotUpdate` condition if the object DNE? We'd only add a `CannotPropagate` condition to the parent?

Reply:
I thought about it, but the argument I'd make is:
* Someone intended `child` to have `superadmin` in it
* `child` is not as intended
* It would be wrong for someone to look at `child` and say "yup, everything looks good."

I know it makes the implementation a bit harder but I think I prefer it the end-user behaviour.

Although now that I think about it, I think we still have a problem after this is fixed - _descendants_ of this namespace might not have `CannotUpdate` cleared. Hmm. I'll file a bug and see if I can reproduce that.

Reply:
Filed #605